### PR TITLE
gprinttyp: a graphical debugging printer for types

### PR DESCRIPTION
**Pull request by Octachron**
This purely internal PR proposes to add a `Gprinttyp` module which can print faithfully type expressions as digraphs using the graphviz format.

This module is intended as a helper for typechecker debugging session (see https://github.com/ocaml/ocaml/commit/71dd0d0043a142fadb880a973354572f9b6dbeef for an example of use case), since graph representation scales better than purely textual representation when there are more than a dozen of nodes involved.

As an illustration, the digraph representation for
```ocaml
val f : (< m : 'a; w : < n : 'a; z : 'b; .. > as 'b; .. > as 'a) -> 'b -> 'a list * 'b list
```
is 
0182-unify-f-l1 4-5

 which is still readable even with the 16 nodes of the type expressions, and the few empty universal type variable binders.
 
 Since printing digraphs on stdout does not work that well when printing hundreds of graphs (my debugging session in #13004 typically produced between 100 and 600 graphs by execution trace), the interface of `Gprinttyp` is oriented toward dumping files in a directory specified by `dump-dirs`, with few helpers functions that helps to provide context.

All feature of the type expressions are covered (type expansions, object fields and polymorphic variants included) and there are few knobs exposed for tuning the level of detail of the graph, and it should be straightforward to add more, in particular for printing levels and scopes.

I imagine that I should mention that the unicode characters in the printers can be removed if required.

### CODE REVIEW

**Comment by gasche** [comment]
This is a nice idea and I agree in principle. I would expect a review for readability (not "correctness", this is debug-only and we don't care if there is a slight issue here or there) before we can merge.

I don't quite follow the actual example (as always): `f` has a function type with two arguments (`... -> ... -> ...`), but I see only one arrow node in your graph. I am also not sure what `, 7` represents, is it the product type? (Then maybe `*` would be more clear.)

**Comment by Octachron** [comment]
Oops, I picked a non-representative subgraph, here is a graph of an unification step with the two arrows present.
0169-unify-f-l1 4-5

The `,` nodes are indeed tuples. I agree that a `x` or `*` would be more readable.

File: typing/gprinttyp.ml
Comment (gasche) [inline typing/gprinttyp.ml]: elide, I think?

File: typing/gprinttyp.ml
Comment (gasche) [inline typing/gprinttyp.ml]: I would include a short comment that the modulo-prime-number stuff is meant to compute a random-looking, regularly spaced colors, in a way that is deterministic on the `id` (and in particular reproducible over several runs of the compiler).

File: typing/gprinttyp.ml
Comment (gasche) [inline typing/gprinttyp.ml]: It would be nicer if `color_id` had its own type `type color_id = Color of int`. This would in particular avoid the temptation of calling `colorize` above on the type node's id.

File: typing/gprinttyp.ml
Comment (gasche) [inline typing/gprinttyp.ml]: I don't understand the logic here -- I'm not sure what `field` is and from a distance I would have expected that the field `baz` of `foo.bar` is `foo.bar.baz`, not `foo.baz`. No big deal, but maybe a comment. Also when reading that I wondered if we should have `Field of t * string` instead of your weird definition, but I'm fine with the weird definition as well.

File: typing/gprinttyp.ml
Comment (gasche) [inline typing/gprinttyp.ml:238]: The name `synth` is not clear. I wonder if the mutable state could be pushed into `name_map`.

File: typing/gprinttyp.ml
Comment (gasche) [inline typing/gprinttyp.ml]: Is the empty string here intentional?

File: typing/gprinttyp.ml
Comment (gasche) [inline typing/gprinttyp.ml]: Nitpick: the code below more consistently uses the `|> numbered [t1; t2]` style.

File: typing/gprinttyp.ml
Comment (gasche) [inline typing/gprinttyp.ml]: Why is  `follow_expansions` not part of `params`?

**Review by gasche — COMMENTED** [review]
I did a quick review (I did not read the dot-producing code in full details). This is the usual nice, slightly over-engineered Florian style and I trust it to perform above expectations at printing graphs from types. Thank you!

I think that you should split the change to `match_row_field` and the `row_field_cell` type to a separate preparatory commit.

I will be happy to approve once you have done a pass taking the review comments into account. (The most invasive one is to mass-rewrite `ellide` into `elide`.)

File: typing/gprinttyp.ml
Comment (Octachron) [reply typing/gprinttyp.ml]: Indeed, fixed.

File: typing/gprinttyp.ml
Comment (Octachron) [reply typing/gprinttyp.ml]: I have added comments about why we are using evenly hue-spaced color for ids.

File: typing/gprinttyp.ml
Comment (Octachron) [reply typing/gprinttyp.ml]: I have moved the `colorize` function inside the `Index` module and made it operate on the graph indices outside of the submodule.

File: typing/gprinttyp.ml
Comment (Octachron) [reply typing/gprinttyp.ml]: I have renamed the constructor to `Named_subnode` to make it clearer that those index are associated to a main index and that the `subnode` of a `subnode` is not really supported at this point.

File: typing/gprinttyp.ml
Comment (Octachron) [reply typing/gprinttyp.ml]: It wasn't, this is fixed.

File: typing/gprinttyp.ml
Comment (Octachron) [reply typing/gprinttyp.ml]: Mostly because I was thinking of it as an internal parameter for the `group` function, but indeed there is no reason to not expose it as a parameter. I have also added an `update_params` function to be able to tweak fields of parameters more easily.

**Review by gasche — APPROVED** [review]
This looks good to me now. I think you could cleanup the history my merging the N-2 middle commits in one.

File: typing/gprinttyp.ml
Comment (lthls) [inline typing/gprinttyp.ml]: Shouldn't that be a kappa (κ or one of the other versions) ?

Comment (Octachron) [reply typing/gprinttyp.ml]: Indeed (even if κ and k are quite confusable depending on the font).

File: typing/gprinttyp.ml
Comment (MisterDA) [inline typing/gprinttyp.ml]:
```suggestion
      | "r" -> "𝜌"
      | "s" -> "𝜎"
      | "p" -> "𝜋"
      | "i" -> "𝜄"
      | "h" -> "𝜂"
      | "k" -> "𝜅"
      | "l" -> "𝜆"
      | "m" -> "𝜇"
```
If I'm not mistaken, you're using Unicode's Mathematical Bold Italic Small symbols, from U+1D736 to U+1D714. The kappa and mu were not exactly consistent…

Comment (Octachron) [reply typing/gprinttyp.ml]: Indeed,  this is better, thanks!

**Comment by Octachron** [comment]
While squashing the history, I have added a basic description of the graph format; since I had been told that the format might be a bit too arcane to decipher without any explanation.

**Comment by gasche** [comment]
@Octachron I suggest that we merge after you fix the conflicts.

**Comment by gasche** [comment]
Thanks for the rebase. CI report:

```
File "/home/runner/work/ocaml/ocaml/typing/rawprinttyp.ml", line 140, characters 23-27:
140 |           match e with None -> fprintf ppf " RFnone"
                             ^^^^
Error: This pattern should not be a constructor, the expected type is
       Types.row_field_cell * Types.row_field option
```

**Comment by Octachron** [comment]
Yep, I fell to the hubris of not checking the rebasing, this should be fixed now.

**Pull request merged**
